connectors/src/connectors/google_drive/temporal/spreadsheets.ts

```diff
@@ -0,0 +1,380 @@
+import { concurrentExecutor } from "@connectors/lib/async_utils";
+import { deleteTable, upsertTableFromCsv } from "@connectors/lib/data_sources";
+import type { GoogleDriveFiles } from "@connectors/lib/models/google_drive";
+import { GoogleDriveSheet } from "@connectors/lib/models/google_drive";
+import { connectorHasAutoPreIngestAllDatabasesFF } from "@connectors/lib/workspace";
+import logger from "@connectors/logger/logger";
+import { ConnectorResource } from "@connectors/resources/connector_resource";
+import type { GoogleDriveObjectType } from "@connectors/types/google_drive";
+
+const MAXIMUM_NUMBER_OF_GSHEET_ROWS = 10000;
+
+type Sheet = sheets_v4.Schema$ValueRange & {
+  id: number;
+  spreadsheet: {
+    id: string;
+    title: string;
+  };
+  title: string;
+};
+
+export function isGoogleDriveSpreadSheetFile(
+  file: GoogleDriveObjectType | GoogleDriveFiles
+) {
+  return file.mimeType === "application/vnd.google-apps.spreadsheet";
+}
+
+function makeTableIdFromSheetId(spreadsheetId: string, sheetId: number) {
+  return `google-spreadsheet-${spreadsheetId}-sheet-${sheetId}`;
+}
+
+async function upsertSheetInDb(connector: ConnectorResource, sheet: Sheet) {
+  await GoogleDriveSheet.upsert({
+    connectorId: connector.id,
+    driveFileId: sheet.spreadsheet.id,
+    driveSheetId: sheet.id,
+    name: sheet.title,
+  });
+}
+
+async function upsertTable(
+  connector: ConnectorResource,
+  sheet: Sheet,
+  rows: string[][]
+) {
+  const dataSourceConfig = await dataSourceConfigFromConnector(connector);
+
+  const { id, spreadsheet, title } = sheet;
+  // Table name will be slugify in front.
+  const tableName = `${spreadsheet.title} - ${title}`;
+  const tableDescription = `Structured data from the Google Spreadsheet (${spreadsheet.title}) and sheet (${title}`;
+
+  const csv = stringify(rows);
+
+  // Upserting is safe: Core truncates any previous table with the same Id before
+  // the operation. Note: Renaming a sheet in Google Drive retains its original Id.
+  await upsertTableFromCsv({
+    dataSourceConfig,
+    tableId: makeTableIdFromSheetId(spreadsheet.id, id),
+    tableName,
+    tableDescription,
+    tableCsv: csv,
+    loggerArgs: {
+      connectorId: connector.id,
+      sheetId: id,
+      spreadsheetId: spreadsheet.id,
+    },
+  });
+}
+
+function makeRandomString(length: number) {
+  return uuidv4.toString().substring(0, length);
+}
+
+function findDataRangeAndSelectRows(allRows: string[][]): string[][] {
+  // Find the first row with data to determine the range.
+  const firstNonEmptyRow = allRows.find((row) =>
+    row.some((cell) => cell.trim() !== "")
+  );
+  if (!firstNonEmptyRow) {
+    return []; // No data found.
+  }
+
+  // Identify the range of data: Start at the first non-empty cell and end at the nearest following empty cell or row end.
+  const startIndex = firstNonEmptyRow.findIndex((cell) => cell.trim() !== "");
+  let endIndex = firstNonEmptyRow.findIndex(
```

Comment:
Ok so if I understand correctly, thanks to this we can't have empty headers 👍

Reply:
Correct.